infrastructure/comms/src/connection/net_address/parser.rs

```diff
@@ -0,0 +1,207 @@
+    }
+
+    /// Parse Onion address
+    pub fn parse_onion(&mut self) -> Option<OnionAddress> {
+        self.read_atomic(|p| {
+            let public_key = match p.read_base32_string() {
+                Some(pk) => {
+                    // Valid onion address lengths
+                    if pk.len() != 16 && pk.len() != 56 {
+                        return None;
+                    }
+
+                    pk
+                },
+                _ => return None,
+            };
+
+            match p.read_until_char(':') {
+                Some(buf) => {
+                    if buf.to_ascii_lowercase() != b".onion" {
+                        return None;
+                    }
+                },
+                None => return None,
+            }
+
+            if p.consume_char(':').is_none() {
+                return None;
+            }
+
+            let port = match p.read_number() {
+                Some(p) => p,
+                None => return None,
+            };
+
+            if port > std::u16::MAX as u64 {
+                return None;
+            }
+
+            Some(OnionAddress { public_key, port: port as u16 })
+        })
+    }
+
+    fn is_base32_char(&self, ch: char) -> bool {
+        if ch >= 'a' && ch <= 'z' {
+            return true;
+        }
+
+        if ch >= '2' && ch <= '7' {
+            return true;
+        }
+
+        false
+    }
+
+    fn read_base32_string(&mut self) -> Option<String> {
+        let mut buf = vec![];
+        while self.pos < self.data.len() {
+            let ch = self.data[self.pos].to_ascii_lowercase();
+            if !self.is_base32_char(ch as char) {
+                break;
+            }
+            buf.push(ch);
+            self.pos += 1;
+        }
+
+        if buf.len() > 0 {
+            match String::from_utf8(buf) {
+                Ok(s) => Some(s),
+                Err(_) => None,
+            }
+        } else {
+            None
+        }
+    }
+
+    fn read_char(&mut self) -> Option<char> {
+        if self.is_end() {
+            return None;
+        } else {
+            let ch = self.data[self.pos];
+            self.pos += 1;
+            Some(ch as char)
+        }
+    }
+
+    fn is_end(&self) -> bool {
+        self.pos == self.data.len()
+    }
+
+    fn consume_char(&mut self, ch: char) -> Option<char> {
+        self.read_char().and_then(|c| if c == ch { Some(ch) } else { None })
+    }
+
+    fn read_number(&mut self) -> Option<u64> {
+        let mut pos = self.pos;
+        let mut number = 0u64;
+        while pos < self.data.len() {
+            let ch = self.data[pos];
+
+            if ch < b'0' || ch > b'9' {
+                break;
+            }
+            number = number * 10u64 + (ch - b'0') as u64;
+            pos += 1;
+        }
+
+        if pos == self.pos {
+            None
+        } else {
+            self.pos = pos;
+            Some(number)
+        }
+    }
+
+    fn read_until_end(&mut self) -> Option<Vec<u8>> {
```

Comment:
Not sure this needs to be an option as a vec can be empty and have length 0.

Reply:
In the case of this method, there's no chance of failure. It's a (debatably pointless) consistency thing, `None` means we couldn't "read until end", Some(Vec) means we could (even if it's empty)